mergekit/scripts/megamerge.py

```diff
@@ -0,0 +1,143 @@
+from mergekit.options import add_merge_options
+
+merges = {}
+
+
+def has_circular_dependency(nodes):
+    def dfs(node, visited, stack):
+        visited[node] = True
+        stack[node] = True
+
+        for dependency in nodes[node]["deps"]:
+            if not visited[dependency]:
+                if dfs(dependency, visited, stack):
+                    return True
+            elif stack[dependency]:
+                return True
+
+        stack[node] = False
+        return False
+
+    visited = {key: False for key in nodes}
+    stack = {key: False for key in nodes}
+
+    for node in nodes:
+        if not visited[node]:
+            if dfs(node, visited, stack):
+                return node
+
+    return None
+
+
+def merge(m, merge_options, force, out_path):
+    # check if output_path exists
+    if os.path.exists(out_path / m):
+        if not force:
+            logging.info("Skipping %s as it already exists", m)
+            del merges[m]
+            return
+        logging.info("Overwriting %s as --force was specified", m)
+
+    if len(merges[m]["deps"]) != 0:
+        for dep in merges[m]["deps"]:
+            if dep in merges:
+                merge(dep, merge_options, force, out_path)
+
+    logging.info("Merging model %s", m)
+    merge_config: MergeConfiguration = MergeConfiguration.model_validate(merges[m])
+    run_merge(
+        merge_config,
+        str(out_path / merges[m]["name"]),
+        options=merge_options,
+    )
+    del merges[m]
+
+
+@click.command("mergekit-mega")
+@click.argument("config_file")
+@click.argument("out_path")
+@click.option(
+    "--verbose", "-v", type=bool, default=False, is_flag=True, help="Verbose logging"
+)
+@click.option(
+    "--force",
+    "-f",
+    type=bool,
+    default=False,
+    is_flag=True,
+    help="overwrite existing merge results instead of skipping them",
+)
+@add_merge_options
+def main(
+    merge_options: MergeOptions,
+    config_file: str,
+    out_path: str,
+    force: bool,
+    verbose: bool,
+):
+    logging.basicConfig(level=logging.INFO if verbose else logging.WARNING)
+
+    out_path = Path(out_path)
+    with open(config_file, "r", encoding="utf-8") as f:
+        data = yaml.load_all(f, Loader=yaml.FullLoader)
+
+        for d in data:
+            if "/" in d["name"]:
+                logging.error("name must not contain a slash")
+                sys.exit(1)
+
+            merges[d["name"]] = d
+            merges[d["name"]]["deps"] = []
+            if "slices" in d:
+                for slc in d["slices"]:
+                    for src in slc["sources"]:
+                        if "model" in src and src["model"] is not None:
+                            model_lora = src["model"].split("+")
+                            # name must not have a slash to avoid path traversal
+                            # therefore, we can use it to check if its a merge from the config
+                            if "/" not in model_lora[0]:
```

Comment:
It's a nice touch that you're allowing previous merges plus a LoRA applied - that didn't occur to me. Might be nice to pull this into a helper function instead of duplicating the logic below though.

Reply:
Yep, good idea, done now